doc/modules/linear_model.rst

```diff
@@ -842,6 +848,98 @@ Ridge Regression`_, see the example below.
+=================================================================
+
+All least-squares linear regression estimators in scikit-learn solve a centered
+version of the problem without intercept and then compute the intercept a
+posteriori from that solution.
+
+This ensures that the ridge solution converges to the minimal norm OLS solution
+when `alpha` tends to zero in the underdeterimined case (more features than
+samples).
+
+The following explains why this approach is valid, both in the penalized and
+unpenalized underdetermined cases.
+
+Intercept of the penalized least-squares solution
+-------------------------------------------------
+
+Recall the definition of the ridge estimator:
+
+.. math::  \min_{w, w_0} { ||X w + w_0 1_s - y||_2 ^ 2 + \alpha ||w||_2 ^ 2}
+
+Here :math:`1_s` is a vector of size `n_samples` where all entries are 1.
+:math:`w_0` is a scalar parameter and is not regularized.
+
+Let's note :math:`\bar{X}` the column vector whose entries the means of the
+columns of :math:`X` and :math:`\bar{y}` the mean scalar value of :math:`y`,
+:math:`X_c` the centered version of :math:`X`, :math:`y_c` the centered version
+of :math:`y` such that:
+
+.. math::  X = X_c + 1_s \bar{X}^{T}  \text{ and } y = y_c + \bar{y} 1_s
+
+Optimizing the original ridge objective with intercept is equivalent to solving
+the following ridge problem on the centered dataset with no intercept:
+
+.. math:: \hat{w} = \min_{w} { ||X_c w - y_c||_2 ^ 2 + \alpha ||w||_2 ^ 2}
+
+and subsequently define the intercept as :math:`\hat{w_0} = \bar{y} -
+\bar{X}^{T} \hat{w}`.
+
+To see this, let's rewrite the data-fit term of the original ridge problem as
+follows (after rearranging the terms and factorizing by :math:`1_s`):
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c +  (\bar{X}^{T} w  + w_0  - \bar{y}) 1_s ||_2^2
+
+Let's define the scalar :math:`\bar{r} = \bar{X}^{T} w  + w_0  - \bar{y}`. We
+can now expand:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c ||_2^2 + ||\bar{r} 1_s ||_2^2 + 2 \bar{r} 1_s^{T} (X_c w - y_c)
+
+If :math:`n` is the number of samples, we have:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 = ||X_c w - y_c ||_2^2 + \bar{r}^2 n + 2 \bar{r} (1_s^{T} X_c w - 1_s^{T} y_c)
+
+Since :math:`X_c` and :math:`y_c` are the centered versions of X and y, we have
+:math:`1_s^{T} X_c = 0` and :math:`1_s^{T} y_c = 0`. Therefore:
+
+.. math::  ||X w + w_0 1_s - y||_2^2 + \alpha ||w||_2^2 = ||X_c w - y_c ||_2^2  + \alpha ||w||_2^2 + \bar{r}^2 n
+
+The first two terms of the right hand side do not depend on :math:`w_0` and are
+minimized by :math:`\hat{w}` the solution of the centered ridge without
+intercept. The third term is positive and therefore subsequently minimized to
+zero by setting :math:`\hat{w_0} = \bar{y} - \bar{X}^{T} \hat{w}`.
+
+Note that the same argument holds for any other penalized linear regression
+estimator (as long as the penalty is such that the solution is unique).
```

Comment:
this explanation is nice and easy to follow.

When I do this in a class I say that at the optimum the gradient wrt w and w_0 should be zero and when you write the gradient wrt w_0 you get directly `\hat{w_0} = \bar{y} - \bar{X}^{T} \hat{w}`